### DISCUSSION TITLE
Packaged successfully on windows, but can only run app from terminal on macOS, not double click.

### DISCUSSION THREAD

**Discussion by JacobBumgarner** [pyinstaller/pyinstaller#6503] [Help] [answered]
I've built an application using various python packages with a PyQt5 frontend. I've been able to freeze the application on windows successfully for distribution, but I cannot figure out how to do it on macOS.

https://github.com/JacobBumgarner/VesselVio

As the title states, I've built it and it can be distributed on Windows, but not macOS. It will not open from a double click on macOS, but it will open from a command line `open dist/VesselVio.app`. To make things even more confusing, after opening it from the command line once, a double click opens it.

So far, I've found that excluding specific imports from an `--add-data:./Library:Library` line allows the app to open from double-click on macOS. Obviously this leads the app to be non-functional, but it tells me that there's something weird going on that can probably be fixed. 

I'm currently building the app using:
```
pyinstaller \
--add-data=./Library:Library \
--additional-hooks-dir=./Hooks \
--icon Icons/icon.icns \
--clean \
--windowed \
VesselVio.py
```

I know that this is a considerable request, but would anyone be willing to help me figure this problem out? If so, I can more carefully detail the app architecture. This is important to me for a publication.

I've spent probably a dozen hours banging my head against this with no success. I've successfully built a distributable for an older version of the app on macOS in 2021, but I cannot figure out what is going wrong. 

Thanks in advance.

**Comment by rokm** [comment, marked as answer]
What version of PyInstaller are you using? The latest (4.7) or the one you have [pinned in your project's requirements (4.3)](https://github.com/JacobBumgarner/VesselVio/blob/9eaa92e16197df193b7659a790a7cb40452278cd/requirements.txt#L26)?

**Reply by JacobBumgarner** [reply]
Thanks for your response. I have tried it with both. 

When I've been trying to build this app, this has been my process: I comment out all imports/functions until the app opens by a double-click. I'll then add some imports/functions back in until it closes, and repeat the process. 

I've currently made it down the rabbit whole to a specific issue: it seems like when two of my library files import one another (circularly?), that's when the application won't open on its own. The script still executes from the terminal by running `python VesselVio.py`, by executing the pyinstaller exe, or by running `open dist/vesselvio.app`. 

Does this give any hint/indication to you as to what might be going on?

**Reply by rokm** [reply]
Hmm, it actually looks like encoding issue to me. Perhaps #6143, because the source .py files that you explicitly collect as part of `Library` directory contain non-ASCII character (the © in copyright notices)? We fixed that in #6146 a while ago, but forgot to back-port it to `v4` branch, so its part of the 4.8 release that was released an hour ago...

Can you try upgrading to 4.8 and rebuilding with that?

**Reply by rokm** [reply]
And if you're actually using PyQt 5.13 that you have pinned (instead of 5.15.x), you will probably need to add 
```python
import os
os.environ["QT_MAC_WANTS_LAYER"] = '1'
```
at the top of VesselVio.py (before the PyQt5 imports) for the application window itself to be shown on Big Sur or later...

**Reply by JacobBumgarner** [reply]
Ok - I'll remove all of the copyright notice ascii characters, will upgrade to 4.8, and will add the `os.environ` line.

Unfortunately the PyQt==5.13 is necessary because of the way that the visualization toolkit interacts with PyQt==15.

I'll run that now and give you and update.

Thanks for taking the time to help with me with this. Fingers crossed...

**Reply by JacobBumgarner** [reply]
These steps worked!! (At least locally). Later tonight when I have access to my old Mac I'm going to send this over and make sure it loads up.

A tentative but huge thank you for now.
